version.js

```diff
@@ -1,92 +1,291 @@
+  // This is considered a "local" version by Python and is not able to be uploaded to PyPI.
+  return `${baseVersion}-alpha.${timestampToUnix(timestamp)}`;
 }
 
-async function getPreviousReleaseVersion(context) {
+/**
+ * @param {string} baseVersion
+ * @param {string} timestamp
+ * @param {string} sha
+ * @returns {string}
+ */
+function localAlphaVersion(baseVersion, timestamp, sha) {
+  return `${baseVersion}-alpha.${timestampToUnix(timestamp)}+${shortHash(sha)}`;
+}
+
+/**
+ * @param {string} ref
+ * @returns {string}
+ */
+function calculateTagVersion(ref) {
+  // Get the version from the tag
+  const tag = ref.replace("refs/tags/", "");
+  localDebug(`tag: ${tag}`);
+  // Ensure it's a valid semver version
+  const parsed = new SemVer(tag);
+  return parsed.version;
+}
+
+/**
+ * @param {string} branchName
+ * @returns {number | undefined}
+ */
+function tryParseVersionBranch(branchName) {
+  const match = branchName.match(/^v(\d+)$/);
+  if (match) {
+    const version = parseInt(match[1], 10);
+    if (!isNaN(version)) {
+      return version;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Checks the PR to see if the major version should be incremented based on labels.
+ * @param {string} commitMessage
+ * @returns {Promise<'minor' | 'major'>}
+ */
+async function getIncrementType(commitMessage) {
+  const prNumber = tryParsePrNumber(commitMessage);
+  if (prNumber === undefined) {
+    return "minor";
+  }
+  const labels = await findAssociatedPrLabels(context.repo, prNumber);
+  return getIncrementTypeFromLabels(labels);
+}
+
+/**
+ * @param {{ name: string }[] | undefined} labels
+ * @returns {'minor' | 'major'}
+ */
+function getIncrementTypeFromLabels(labels) {
+  if (hasNeedsMajorReleaseLabel(labels)) {
+    return "major";
+  }
+  return "minor";
+}
+
+/**
+ * @param {string} commitMessage
+ * @returns {number | undefined}
+ */
+function tryParsePrNumber(commitMessage) {
+  const prMatch = commitMessage.match(/\(#(\d+)\)/);
+  if (prMatch) {
+    const num = parseInt(prMatch[1], 10);
+    if (!isNaN(num)) {
+      return num;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * @param {{ owner:string, repo: string }} repo
+ * @param {number} prNumber
+ * @returns {Promise<{ name: string }[] | undefined>}
+ */
+async function findAssociatedPrLabels(repo, prNumber) {
+  const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });
+  const pr = await octokit.rest.pulls.get({
+    ...repo,
+    pull_number: prNumber,
+  });
+  return pr.data?.labels;
+}
+
+/**
+ * @param {{ name: string }[] | undefined} labels
+ * @returns {boolean}
+ */
+function hasNeedsMajorReleaseLabel(labels) {
+  if (!labels) {
+    return false;
+  }
+  localDebug(`PR labels: ${labels.map((label) => label.name).join(", ")}`);
+  return labels.some((label) => label.name === "needs-release/major");
+}
+
+/**
+ * Check if we're building, or merging to, a version branch.
```

Comment:
Not necessarily for this PR, but can we add some automation that creates the version branches?

E.g. when we cut a new major release, we could create a version branch with the previous state of the trunk

Reply:
Yeah, that'd be neat. That wouldn't be in this action though - that would be part of the release workflow.